src/lyg.rs

```diff
@@ -0,0 +1,326 @@
+        };
+        ({ ! $pat:tt }) => {
+            negative_lookahead($pat)
+        };
+        ({ ! $start:tt ..= $end:tt }) => {
+            negative_lookahead($start..=$end)
+        };
+        ($rule:ident) => {
+            call(stringify!($rule))
+        };
+        ({ $name:ident : $rule:tt }) => {
+            rule!($rule).field(stringify!($name))
+        };
+        ({ $rule:tt ? }) => {
+            rule!($rule).opt()
+        };
+        ({ $elem:tt * }) => {
+            rule!($elem).repeat_many()
+        };
+        ({ $elem:tt + }) => {
+            rule!($elem).repeat_more()
+        };
+        ({ $elem:tt + % $sep:tt }) => {
+            rule!($elem).repeat_more_sep(rule!($sep), SepKind::Simple)
+        };
+        ({ $rule0:tt $(| $rule:tt)+ }) => {
+            rule!($rule0) $(| rule!($rule))+
+        };
+        ({ $rule0:tt $($rule:tt)* }) => {
+            rule!($rule0) $(+ rule!($rule))*
+        };
+        ($pat:expr) => {
+            eat($pat)
+        };
+    }
+
+    macro_rules! grammar {
+        ($($rule_name:ident = $($rule:tt)|+;)*) => ({
+            let mut grammar = Grammar::new();
+            $(grammar.define(
+                cx.intern(stringify!($rule_name)),
+                rule!({ $($rule)|+ }).finish(cx),
+            );)*
+            grammar
+        })
+    }
+
+    // Main grammar.
+    let mut grammar = grammar! {
+        Grammar = { FileStart {rules:{RuleDef*}} FileEnd };
+        RuleDef = { {name:Ident} "=" {rule:Or} ";" };
+        Or = {{"|"?} {rules:{Concat+ % "|"}}};
+        Concat = {rules:{Rule+}};
+        Rule = { {{ {field:Ident} ":" }?} {rule:Primary} {{modifier:Modifier}?} };
+        Primary =
+            {Eat:Pattern} |
+            {Call:Ident} |
+            {Group:{ "{" {{or:Or}?} "}" }};
+        Modifier =
+            {Opt:"?"} |
+            {Repeat:{ {repeat:Repeat} {{ {kind:SepKind} {sep:Primary} }?} }};
+        Repeat =
+            {Many:"*"} |
+            {More:"+"};
+        SepKind =
+            {Simple:"%"} |
+            // HACK(eddyb) should be "%%", but `rustc`'s `proc_macro` server doesn't
+            // always preserve jointness, except within multi-character Rust operators.
+            {Trailing:{"%" "%"}};
+        Pattern =
+            {Str:StrLit} |
+            {CharRange:{ {{start:CharLit}?} ".." {{end:CharLit}?} }} |
+            {CharRangeInclusive:{ {{start:CharLit}?} "..=" {end:CharLit} }};
+    };
+
+    // Lexical fragment of the grammar.
+    grammar.extend(grammar! {
+        FileStart = "";
+        FileEnd = "";
+
+        Ident = IDENT;
+
+        // FIXME(eddyb) restrict literals, once `proc_macro` allows it.
+        StrLit = LITERAL;
+        CharLit = LITERAL;
+    });
+
+    grammar
+}
+
+type Handle<'a, 'i> = crate::slow_bruteforce_interpreter::Handle<'a, 'i, PMPat, TokenStream>;
+type RecordHandle<'a, 'i> =
+    crate::slow_bruteforce_interpreter::RecordHandle<'a, 'i, PMPat, TokenStream>;
+
+macro_rules! rec {
```

Comment:
Should this be exposed somewhere? It seems very useful for navigating a tree.

Reply:
Yeah I think I should expose it from the interpreter module.
Something that might not be obvious, is that I'm not sure the current design for handles, especially the ones dealing with records, is a dead-end.

I think something more focused on disambiguation might go a long way, but I'm not sure yet how I want to play that.

Probably the easy thing would be to deal with fields in a very lazy manner, and only expand the parse forest as much as needed to get at the fields.

So you could end up with e.g. `Expr({Binary:{...}|{...}})` instead of `Expr({Binary:{...}}) | Expr({Binary:{...}})`, when printing an `Expr` handle.

But disambiguation would be bottom-up and so every handle you'd get in your disambiguation logic, would be fully disambiguated if you were to recurse on it.